File: ios-native/Dimo/Email/Gmail/GmailOAuthClient.swift
import AuthenticationServices
import Foundation
import UIKit

enum GmailURLSession {
  static func make() -> URLSession {
    let configuration = URLSessionConfiguration.ephemeral
    configuration.urlCache = nil
    configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
    configuration.httpCookieStorage = nil
    configuration.httpShouldSetCookies = false
    configuration.waitsForConnectivity = true
    return URLSession(configuration: configuration)
  }
}

struct GmailOAuthConfiguration: Hashable, Sendable {
  static let authorizationEndpoint = URL(string: "https://accounts.google.com/o/oauth2/v2/auth")!
  static let tokenEndpoint = URL(string: "https://oauth2.googleapis.com/token")!
  static let revocationEndpoint = URL(string: "https://oauth2.googleapis.com/revoke")!
  static let userInfoEndpoint = URL(string: "https://openidconnect.googleapis.com/v1/userinfo")!
  static let readOnlyScope = "https://www.googleapis.com/auth/gmail.readonly"

  var clientId: String
  var redirectScheme: String
  var redirectURI: String

  init(clientId: String, redirectScheme: String, redirectURI: String? = nil) {
    self.clientId = clientId
    self.redirectScheme = redirectScheme
    self.redirectURI = redirectURI ?? "\(redirectScheme):/oauthredirect"
  }

  static func fromAppConfig() throws -> GmailOAuthConfiguration {
    guard AppConfig.isGmailConfigured else {
      throw GmailOAuthError.notConfigured
    }
    return GmailOAuthConfiguration(
      clientId: AppConfig.gmailOAuthClientID,
      redirectScheme: AppConfig.gmailOAuthRedirectScheme,
      redirectURI: AppConfig.gmailOAuthRedirectURI
    )
  }
}

struct GmailAccessToken: Hashable, Sendable {
  var value: String
  var expiresAt: Date
}

protocol GmailAccessTokenProviding: Sendable {
  func accessToken(subject: String, dimoUserId: String, forceRefresh: Bool) async throws
    -> GmailAccessToken
  func invalidate(subject: String) async
}

/// Performs Gmail's separate installed-app OAuth flow. It never uses the WorkOS token.
final class GmailOAuthClient: NSObject, @unchecked Sendable {
  private let configuration: GmailOAuthConfiguration
  private let vault: GmailCredentialVault
  private let session: URLSession
  @MainActor private var webAuthenticationSession: ASWebAuthenticationSession?

  init(
    configuration: GmailOAuthConfiguration,
    vault: GmailCredentialVault,
    session: URLSession = GmailURLSession.make()
  ) {
    self.configuration = configuration
    self.vault = vault
    self.session = session
  }

  func connect(dimoUserId: String) async throws -> ConnectedGmailAccount {
    let verifier = PKCE.makeVerifier()
    let state = PKCE.makeState()
    let authorizationURL = try makeAuthorizationURL(verifier: verifier, state: state)
    let callback = try await startWebAuthentication(url: authorizationURL)
    let queryItems = URLComponents(url: callback, resolvingAgainstBaseURL: false)?.queryItems ?? []
    var parameters: [String: String] = [:]
    for item in queryItems {
      if let value = item.value { parameters[item.name] = value }
    }
    if let oauthError = parameters["error"] {
      if oauthError == "access_denied" { throw GmailOAuthError.cancelled }
      throw GmailOAuthError.authorization(oauthError)
    }
    guard parameters["state"] == state else { throw GmailOAuthError.stateMismatch }
    guard let code = parameters["code"], !code.isEmpty else { throw GmailOAuthError.missingCode }

    let token = try await exchangeCode(code, verifier: verifier)
    guard let refreshToken = token.refreshToken, !refreshToken.isEmpty else {
      throw GmailOAuthError.missingRefreshToken
    }
    let identity = try await fetchIdentity(accessToken: token.accessToken)
    let credential = GmailStoredCredential(
      subject: identity.subject,
      emailAddress: identity.emailAddress,
      refreshToken: refreshToken,
      connectedAt: .now
    )
    try await vault.upsert(credential, dimoUserId: dimoUserId)
    return credential.account
  }

  @MainActor
  func cancel() {
    webAuthenticationSession?.cancel()
    webAuthenticationSession = nil
  }

  /// Revocation is best effort; local credentials are always removed.
  func disconnect(subject: String, dimoUserId: String) async throws {
    let refreshToken = try await vault.credential(subject: subject, dimoUserId: dimoUserId)?.refreshToken
    if let refreshToken { try? await revoke(token: refreshToken) }
    try await vault.remove(subject: subject, dimoUserId: dimoUserId)
  }

  func deleteLocalCredentialsOnSignOut(dimoUserId: String) async {
    try? await vault.removeAll(dimoUserId: dimoUserId)
  }

  private func makeAuthorizationURL(verifier: String, state: String) throws -> URL {
    var components = URLComponents(
      url: GmailOAuthConfiguration.authorizationEndpoint,
      resolvingAgainstBaseURL: false
    )!
    components.queryItems = [
      URLQueryItem(name: "client_id", value: configuration.clientId),
      URLQueryItem(name: "redirect_uri", value: configuration.redirectURI),
      URLQueryItem(name: "response_type", value: "code"),
      URLQueryItem(
        name: "scope",
        value: ["openid", "email", GmailOAuthConfiguration.readOnlyScope].joined(separator: " ")
      ),
      URLQueryItem(name: "access_type", value: "offline"),
      URLQueryItem(name: "prompt", value: "consent select_account"),
      URLQueryItem(name: "include_granted_scopes", value: "true"),
      URLQueryItem(name: "code_challenge", value: PKCE.challenge(for: verifier)),
      URLQueryItem(name: "code_challenge_method", value: "S256"),
      URLQueryItem(name: "state", value: state),
    ]
    guard let url = components.url else { throw GmailOAuthError.invalidAuthorizationURL }
    return url
  }

  private func exchangeCode(_ code: String, verifier: String) async throws -> OAuthTokenResponse {
    try await postForm(
      to: GmailOAuthConfiguration.tokenEndpoint,
      values: [
        "client_id": configuration.clientId,
        "code": code,
        "code_verifier": verifier,
        "grant_type": "authorization_code",
        "redirect_uri": configuration.redirectURI,
      ],
      as: OAuthTokenResponse.self
    )
  }

  private func fetchIdentity(accessToken: String) async throws -> GmailIdentityResponse {
    var request = URLRequest(url: GmailOAuthConfiguration.userInfoEndpoint)
    request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
    let (data, response) = try await session.data(for: request)
    try validate(response: response, data: data)
    do {
      return try JSONDecoder().decode(GmailIdentityResponse.self, from: data)
    } catch {
      throw GmailOAuthError.invalidResponse
    }
  }

  private func revoke(token: String) async throws {
    var components = URLComponents(
      url: GmailOAuthConfiguration.revocationEndpoint,
      resolvingAgainstBaseURL: false
    )!
    components.queryItems = [URLQueryItem(name: "token", value: token)]
    var request = URLRequest(url: components.url!)
    request.httpMethod = "POST"
    request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
    let (data, response) = try await session.data(for: request)
    try validate(response: response, data: data)
  }

  private func postForm<Response: Decodable>(
    to url: URL,
    values: [String: String],
    as type: Response.Type
  ) async throws -> Response {
    var body = URLComponents()
    body.queryItems = values.sorted { $0.key < $1.key }.map(URLQueryItem.init(name:value:))
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
    request.httpBody = body.percentEncodedQuery?.data(using: .utf8)
    let (data, response) = try await session.data(for: request)
    try validate(response: response, data: data)
    let decoder = JSONDecoder()
    decoder.keyDecodingStrategy = .convertFromSnakeCase
    do {
      return try decoder.decode(type, from: data)
    } catch {
      throw GmailOAuthError.invalidResponse
    }
  }

  private func validate(response: URLResponse, data: Data) throws {
    guard let http = response as? HTTPURLResponse else { throw GmailOAuthError.invalidResponse }
    guard (200..<300).contains(http.statusCode) else {
      let payload = try? JSONDecoder().decode(GoogleOAuthErrorResponse.self, from: data)
      let message = payload?.errorDescription ?? payload?.error ?? "HTTP \(http.statusCode)"
      if payload?.error == "invalid_grant" { throw GmailOAuthError.requiresReconnect }
      throw GmailOAuthError.server(message)
    }
  }

  @MainActor
  private func startWebAuthentication(url: URL) async throws -> URL {
    try await withCheckedThrowingContinuation { continuation in
      let authSession = ASWebAuthenticationSession(
        url: url,
        callbackURLScheme: configuration.redirectScheme
      ) { callbackURL, error in
        Task { @MainActor [weak self] in self?.webAuthenticationSession = nil }
        if let error {
          let nsError = error as NSError
          if nsError.domain == ASWebAuthenticationSessionErrorDomain,
             nsError.code == ASWebAuthenticationSessionError.canceledLogin.rawValue {
            continuation.resume(throwing: GmailOAuthError.cancelled)
          } else {
            continuation.resume(throwing: error)
          }
          return
        }
        guard let callbackURL else {
          continuation.resume(throwing: GmailOAuthError.missingCode)
          return
        }
        continuation.resume(returning: callbackURL)
      }
      authSession.presentationContextProvider = self
      authSession.prefersEphemeralWebBrowserSession = false
      webAuthenticationSession = authSession
      guard authSession.start() else {
        webAuthenticationSession = nil
        continuation.resume(throwing: GmailOAuthError.couldNotStartSession)
        return
      }
    }
  }
}

extension GmailOAuthClient: ASWebAuthenticationPresentationContextProviding {
  func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
    let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
    let windows = scenes.flatMap(\.windows)
    if let window = windows.first(where: \.isKeyWindow) ?? windows.first { return window }
    guard let scene = scenes.first else {
      preconditionFailure("Gmail authentication requires an active window scene")
    }
    return UIWindow(windowScene: scene)
  }
}

actor GmailAccessTokenManager: GmailAccessTokenProviding {
  private let configuration: GmailOAuthConfiguration
  private let vault: GmailCredentialVault
  private let session: URLSession
  private var cachedTokens: [String: GmailAccessToken] = [:]

  init(
    configuration: GmailOAuthConfiguration,
    vault: GmailCredentialVault,
    session: URLSession = GmailURLSession.make()
  ) {
    self.configuration = configuration
    self.vault = vault
    self.session = session
  }

  func accessToken(
    subject: String,
    dimoUserId: String,
    forceRefresh: Bool = false
  ) async throws -> GmailAccessToken {
    if !forceRefresh, let cached = cachedTokens[subject], cached.expiresAt.timeIntervalSinceNow > 60 {
      return cached
    }
    guard let credential = try await vault.credential(subject: subject, dimoUserId: dimoUserId) else {
      throw GmailOAuthError.requiresReconnect
    }
    var form = URLComponents()
    form.queryItems = [
      URLQueryItem(name: "client_id", value: configuration.clientId),
      URLQueryItem(name: "refresh_token", value: credential.refreshToken),
      URLQueryItem(name: "grant_type", value: "refresh_token"),
    ]
    var request = URLRequest(url: GmailOAuthConfiguration.tokenEndpoint)
    request.httpMethod = "POST"
    request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
    request.httpBody = form.percentEncodedQuery?.data(using: .utf8)
    let (data, response) = try await session.data(for: request)
    guard let http = response as? HTTPURLResponse else { throw GmailOAuthError.invalidResponse }
    guard (200..<300).contains(http.statusCode) else {
      let payload = try? JSONDecoder().decode(GoogleOAuthErrorResponse.self, from: data)
      if payload?.error == "invalid_grant" { throw GmailOAuthError.requiresReconnect }
      throw GmailOAuthError.server(payload?.errorDescription ?? payload?.error ?? "HTTP \(http.statusCode)")
    }
    let decoder = JSONDecoder()
    decoder.keyDecodingStrategy = .convertFromSnakeCase
    guard let response = try? decoder.decode(OAuthTokenResponse.self, from: data) else {
      throw GmailOAuthError.invalidResponse
    }
    let token = GmailAccessToken(
      value: response.accessToken,
      expiresAt: Date().addingTimeInterval(TimeInterval(max(response.expiresIn - 60, 1)))
    )
    cachedTokens[subject] = token
    return token
  }

  func invalidate(subject: String) {
    cachedTokens[subject] = nil
  }

  func clearAll() {
    cachedTokens.removeAll()
  }
}

private struct OAuthTokenResponse: Decodable {
  var accessToken: String
  var expiresIn: Int
  var refreshToken: String?
}

private struct GmailIdentityResponse: Decodable {
  var subject: String
  var emailAddress: String

  enum CodingKeys: String, CodingKey {
    case subject = "sub"
    case emailAddress = "email"
  }
}

private struct GoogleOAuthErrorResponse: Decodable {
  var error: String?
  var errorDescription: String?

  enum CodingKeys: String, CodingKey {
    case error
    case errorDescription = "error_description"
  }
}

enum GmailOAuthError: LocalizedError, Sendable {
  case notConfigured
  case invalidAuthorizationURL
  case couldNotStartSession
  case cancelled
  case stateMismatch
  case missingCode
  case missingRefreshToken
  case invalidResponse
  case requiresReconnect
  case authorization(String)
  case server(String)

  var errorDescription: String? {
    switch self {
    case .notConfigured: return "Gmail OAuth is not configured."
    case .invalidAuthorizationURL: return "The Gmail authorization URL is invalid."
    case .couldNotStartSession: return "Gmail sign-in could not be opened."
    case .cancelled: return "Gmail sign-in was cancelled."
    case .stateMismatch: return "Gmail sign-in failed its security check."
    case .missingCode: return "Gmail did not return an authorization code."
    case .missingRefreshToken: return "Gmail did not grant offline access. Please connect again."
    case .invalidResponse: return "Gmail returned an invalid response."
    case .requiresReconnect: return "Gmail access expired or was revoked. Please connect again."
    case .authorization(let message): return "Gmail authorization failed: \(message)"
    case .server(let message): return "Gmail authorization failed: \(message)"
    }
  }
}

File: android-native/app/src/main/java/app/dimo/android/email/gmail/GmailOAuthClient.kt
package app.dimo.android.email.gmail

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import app.dimo.android.app.AppConfig
import app.dimo.android.auth.PKCE
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.FormBody
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request

/**
 * Port of `ios-native/Dimo/Email/Gmail/GmailOAuthClient.swift`.
 *
 * Gmail has its own installed-app OAuth flow and never uses the WorkOS token. The
 * browser leg is a Custom Tab instead of `ASWebAuthenticationSession`; PKCE, the
 * `state` round-trip and the subject check on reconnect are identical.
 */
object GmailOAuthConfigurationDefaults {
  const val AUTHORIZATION_ENDPOINT = "https://accounts.google.com/o/oauth2/v2/auth"
  const val TOKEN_ENDPOINT = "https://oauth2.googleapis.com/token"
  const val REVOCATION_ENDPOINT = "https://oauth2.googleapis.com/revoke"
  const val USER_INFO_ENDPOINT = "https://openidconnect.googleapis.com/v1/userinfo"
  const val READ_ONLY_SCOPE = "https://www.googleapis.com/auth/gmail.readonly"
}

data class GmailOAuthConfiguration(
  val clientId: String,
  val redirectScheme: String,
  val redirectUri: String = "$redirectScheme:/oauthredirect",
) {
  companion object {
    fun fromAppConfig(): GmailOAuthConfiguration {
      if (!AppConfig.isGmailConfigured) throw GmailOAuthException.NotConfigured
      return GmailOAuthConfiguration(
        clientId = AppConfig.gmailOAuthClientID,
        redirectScheme = AppConfig.gmailOAuthRedirectScheme,
        redirectUri = AppConfig.gmailOAuthRedirectURI,
      )
    }
  }
}

data class GmailAccessToken(val value: String, val expiresAt: Long)

interface GmailAccessTokenProviding {
  suspend fun accessToken(
    subject: String,
    dimoUserId: String,
    forceRefresh: Boolean = false,
  ): GmailAccessToken

  suspend fun invalidate(subject: String)
}

/**
 * Bridges the Gmail OAuth redirect back to the suspended connect call.
 *
 * Kept separate from `AuthRedirectBus` so a Gmail redirect can never complete a
 * parked WorkOS sign-in, or the reverse — they use different URI schemes and can
 * in principle be in flight at once.
 */
object GmailRedirectBus {
  @Volatile
  private var pending: CompletableDeferred<Uri>? = null

  fun expectRedirect(): CompletableDeferred<Uri> {
    pending?.cancel()
    return CompletableDeferred<Uri>().also { pending = it }
  }

  fun publish(uri: Uri): Boolean {
    val deferred = pending ?: return false
    pending = null
    return deferred.complete(uri)
  }

  fun cancelPending() {
    pending?.completeExceptionally(GmailOAuthException.Cancelled)
    pending = null
  }

  fun hasPending(): Boolean = pending?.isActive == true
}

class GmailOAuthClient(
  private val configuration: GmailOAuthConfiguration,
  private val vault: GmailCredentialVault,
  private val http: OkHttpClient = GmailHttp.client,
) {
  private val json = Json { ignoreUnknownKeys = true }

  suspend fun connect(context: Context, dimoUserId: String): ConnectedGmailAccount =
    authorize(context, dimoUserId, loginHint = null, expectedSubject = null)

  /**
   * Replaces the stored refresh token for an existing account without wiping local
   * email rows. Verifies Google returns the same subject.
   */
  suspend fun reauthorize(
    context: Context,
    subject: String,
    emailAddress: String,
    dimoUserId: String,
  ): ConnectedGmailAccount {
    val existing = vault.credential(subject, dimoUserId)
    return authorize(
      context = context,
      dimoUserId = dimoUserId,
      loginHint = emailAddress,
      expectedSubject = subject,
      connectedAt = existing?.connectedAt,
    )
  }

  /** Revocation is best effort; local credentials are always removed. */
  suspend fun disconnect(subject: String, dimoUserId: String) {
    val refreshToken = vault.credential(subject, dimoUserId)?.refreshToken
    if (refreshToken != null) runCatching { revoke(refreshToken) }
    vault.remove(subject, dimoUserId)
  }

  suspend fun deleteLocalCredentialsOnSignOut(dimoUserId: String) {
    runCatching { vault.removeAll(dimoUserId) }
  }

  // MARK: - Private

  private suspend fun authorize(
    context: Context,
    dimoUserId: String,
    loginHint: String?,
    expectedSubject: String?,
    connectedAt: Long? = null,
  ): ConnectedGmailAccount {
    val verifier = PKCE.makeVerifier()
    val state = PKCE.makeState()
    val url = authorizationUrl(verifier, state, loginHint)

    val redirect = GmailRedirectBus.expectRedirect()
    launchCustomTab(context, url)
    val callback = redirect.await()

    callback.getQueryParameter("error")?.let { oauthError ->
      if (oauthError == "access_denied") throw GmailOAuthException.Cancelled
      throw GmailOAuthException.Authorization(oauthError)
    }
    if (callback.getQueryParameter("state") != state) throw GmailOAuthException.StateMismatch
    val code = callback.getQueryParameter("code")?.takeIf { it.isNotEmpty() }
      ?: throw GmailOAuthException.MissingCode

    val token = exchangeCode(code, verifier)
    val refreshToken = token.refreshToken?.takeIf { it.isNotEmpty() }
      ?: throw GmailOAuthException.MissingRefreshToken
    val identity = fetchIdentity(token.accessToken)
    if (expectedSubject != null && identity.subject != expectedSubject) {
      throw GmailOAuthException.AccountMismatch
    }
    val credential = GmailStoredCredential(
      subject = identity.subject,
      emailAddress = identity.emailAddress,
      refreshToken = refreshToken,
      connectedAt = connectedAt ?: System.currentTimeMillis(),
    )
    vault.upsert(credential, dimoUserId)
    return credential.account
  }

  private fun authorizationUrl(verifier: String, state: String, loginHint: String?): Uri {
    val builder = Uri.parse(GmailOAuthConfigurationDefaults.AUTHORIZATION_ENDPOINT)
      .buildUpon()
      .appendQueryParameter("client_id", configuration.clientId)
      .appendQueryParameter("redirect_uri", configuration.redirectUri)
      .appendQueryParameter("response_type", "code")
      .appendQueryParameter(
        "scope",
        listOf("openid", "email", GmailOAuthConfigurationDefaults.READ_ONLY_SCOPE)
          .joinToString(" "),
      )
      // `offline` + a forced consent screen is what makes Google return a refresh
      // token; without it a reconnect silently yields access-token-only grants.
      .appendQueryParameter("access_type", "offline")
      .appendQueryParameter("prompt", "consent select_account")
      .appendQueryParameter("include_granted_scopes", "true")
      .appendQueryParameter("code_challenge", PKCE.challenge(verifier))
      .appendQueryParameter("code_challenge_method", "S256")
      .appendQueryParameter("state", state)
    loginHint?.trim()?.takeIf { it.isNotEmpty() }?.let {
      builder.appendQueryParameter("login_hint", it)
    }
    return builder.build()
  }

  private fun launchCustomTab(context: Context, url: Uri) {
    val intent = CustomTabsIntent.Builder()
      .setShowTitle(false)
      .setUrlBarHidingEnabled(true)
      .build()
    intent.intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
    intent.launchUrl(context, url)
  }

  private suspend fun exchangeCode(code: String, verifier: String): OAuthTokenResponse =
    postForm(
      GmailOAuthConfigurationDefaults.TOKEN_ENDPOINT,
      mapOf(
        "client_id" to configuration.clientId,
        "code" to code,
        "code_verifier" to verifier,
        "grant_type" to "authorization_code",
        "redirect_uri" to configuration.redirectUri,
      ),
    )

  private suspend fun fetchIdentity(accessToken: String): GmailIdentityResponse =
    withContext(Dispatchers.IO) {
      val request = Request.Builder()
        .url(GmailOAuthConfigurationDefaults.USER_INFO_ENDPOINT)
        .header("Authorization", "Bearer $accessToken")
        .build()
      http.newCall(request).execute().use { response ->
        val body = response.body.string()
        validate(response.code, body)
        runCatching { json.decodeFromString<GmailIdentityResponse>(body) }.getOrNull()
          ?: throw GmailOAuthException.InvalidResponse
      }
    }

  private suspend fun revoke(token: String) = withContext(Dispatchers.IO) {
    val url = GmailOAuthConfigurationDefaults.REVOCATION_ENDPOINT.toHttpUrl()
      .newBuilder()
      .addQueryParameter("token", token)
      .build()
    val request = Request.Builder()
      .url(url)
      .post(FormBody.Builder().build())
      .build()
    http.newCall(request).execute().use { response ->
      validate(response.code, response.body.string())
    }
  }

  private suspend fun postForm(url: String, values: Map<String, String>): OAuthTokenResponse =
    withContext(Dispatchers.IO) {
      val form = FormBody.Builder().apply {
        values.toSortedMap().forEach { (key, value) -> add(key, value) }
      }.build()
      val request = Request.Builder().url(url).post(form).build()
      http.newCall(request).execute().use { response ->
        val body = response.body.string()
        validate(response.code, body)
        runCatching { json.decodeFromString<OAuthTokenResponse>(body) }.getOrNull()
          ?: throw GmailOAuthException.InvalidResponse
      }
    }

  private fun validate(code: Int, body: String) {
    if (code in 200..299) return
    val payload = runCatching { json.decodeFromString<GoogleOAuthErrorResponse>(body) }.getOrNull()
    // `invalid_grant` is the one error the user can fix, by reconnecting.
    if (payload?.error == "invalid_grant") throw GmailOAuthException.RequiresReconnect
    throw GmailOAuthException.Server(
      payload?.errorDescription ?: payload?.error ?: "HTTP $code",
    )
  }
}

/**
 * Exchanges refresh tokens for short-lived access tokens and caches them in
 * memory. Port of the `GmailAccessTokenManager` actor.
 */
class GmailAccessTokenManager(
  private val configuration: GmailOAuthConfiguration,
  private val vault: GmailCredentialVault,
  private val http: OkHttpClient = GmailHttp.client,
) : GmailAccessTokenProviding {
  private val json = Json { ignoreUnknownKeys = true }
  private val cached = ConcurrentHashMap<String, GmailAccessToken>()

  /** Serializes refreshes so a burst of Gmail calls cannot spend the same grant twice. */
  private val mutex = Mutex()

  override suspend fun accessToken(
    subject: String,
    dimoUserId: String,
    forceRefresh: Boolean,
  ): GmailAccessToken = mutex.withLock {
    if (!forceRefresh) {
      val hit = cached[subject]
      if (hit != null && hit.expiresAt - System.currentTimeMillis() > MIN_REMAINING_MS) return hit
    }
    val credential = vault.credential(subject, dimoUserId)
      ?: throw GmailOAuthException.RequiresReconnect
    val response = withContext(Dispatchers.IO) {
      val form = FormBody.Builder()
        .add("client_id", configuration.clientId)
        .add("refresh_token", credential.refreshToken)
        .add("grant_type", "refresh_token")
        .build()
      val request = Request.Builder()
        .url(GmailOAuthConfigurationDefaults.TOKEN_ENDPOINT)
        .post(form)
        .build()
      http.newCall(request).execute().use { httpResponse ->
        val body = httpResponse.body.string()
        if (httpResponse.code !in 200..299) {
          val payload = runCatching {
            json.decodeFromString<GoogleOAuthErrorResponse>(body)
          }.getOrNull()
          if (payload?.error == "invalid_grant") throw GmailOAuthException.RequiresReconnect
          throw GmailOAuthException.Server(
            payload?.errorDescription ?: payload?.error ?: "HTTP ${httpResponse.code}",
          )
        }
        runCatching { json.decodeFromString<OAuthTokenResponse>(body) }.getOrNull()
          ?: throw GmailOAuthException.InvalidResponse
      }
    }
    // Expire a minute early so a token cannot lapse mid-request.
    val token = GmailAccessToken(
      value = response.accessToken,
      expiresAt = System.currentTimeMillis() +
        maxOf(response.expiresIn - 60, 1) * 1000L,
    )
    cached[subject] = token
    token
  }

  override suspend fun invalidate(subject: String) {
    cached.remove(subject)
  }

  fun clearAll() = cached.clear()

  private companion object {
    const val MIN_REMAINING_MS = 60_000L
  }
}

/** Shared, cookie-free OkHttp client for every Google call. */
object GmailHttp {
  val client: OkHttpClient by lazy {
    OkHttpClient.Builder()
      .retryOnConnectionFailure(true)
      .build()
  }
}

@Serializable
private data class OAuthTokenResponse(
  @SerialName("access_token") val accessToken: String,
  @SerialName("expires_in") val expiresIn: Int,
  @SerialName("refresh_token") val refreshToken: String? = null,
)

@Serializable
private data class GmailIdentityResponse(
  @SerialName("sub") val subject: String,
  @SerialName("email") val emailAddress: String,
)

@Serializable
private data class GoogleOAuthErrorResponse(
  val error: String? = null,
  @SerialName("error_description") val errorDescription: String? = null,
)

sealed class GmailOAuthException(message: String) : Exception(message) {
  object NotConfigured : GmailOAuthException("Gmail OAuth is not configured.")
  object Cancelled : GmailOAuthException("Gmail sign-in was cancelled.")
  object StateMismatch : GmailOAuthException("Gmail sign-in failed its security check.")
  object MissingCode : GmailOAuthException("Gmail did not return an authorization code.")
  object MissingRefreshToken : GmailOAuthException(
    "Gmail did not grant offline access. Reconnect this account and approve access again.",
  )
  object InvalidResponse : GmailOAuthException("Gmail returned an invalid response.")
  object RequiresReconnect : GmailOAuthException(
    "Gmail access expired or was revoked. Reconnect this account to continue.",
  )
  object AccountMismatch : GmailOAuthException(
    "That Google account does not match the connected Gmail inbox. " +
      "Choose the same account to reconnect.",
  )

  class Authorization(reason: String) : GmailOAuthException("Gmail authorization failed: $reason")
  class Server(reason: String) : GmailOAuthException("Gmail authorization failed: $reason")
}
